Exception while handling click event in inventory

`[23:24:13 ERROR]: [BellClaims] Exception while handling click event in inventory 'Claim Creation', slot=4, item=BELL
java.lang.NullPointerException: Cannot invoke "java.lang.Class.getConstructor(java.lang.Class[])" because the return value of "java.util.EnumMap.get(Object)" is null
        at BellClaims-0.4.4.jar/com.github.stefvanschie.inventoryframework.util.version.VersionMatcher.newAnvilInventory(VersionMatcher.java:69) ~[BellClaims-0.4.4.jar:?]
        at BellClaims-0.4.4.jar/com.github.stefvanschie.inventoryframework.gui.type.AnvilGui.<init>(AnvilGui.java:78) ~[BellClaims-0.4.4.jar:?]
        at BellClaims-0.4.4.jar/dev.mizarc.bellclaims.interaction.menus.management.ClaimNamingMenu.open(ClaimNamingMenu.kt:36) ~[BellClaims-0.4.4.jar:?]
        at BellClaims-0.4.4.jar/dev.mizarc.bellclaims.interaction.menus.MenuNavigator.openMenu(MenuNavigator.kt:18) ~[BellClaims-0.4.4.jar:?]
        at BellClaims-0.4.4.jar/dev.mizarc.bellclaims.interaction.menus.management.ClaimCreationMenu.open$lambda$3(ClaimCreationMenu.kt:63) ~[BellClaims-0.4.4.jar:?]
        at BellClaims-0.4.4.jar/com.github.stefvanschie.inventoryframework.gui.GuiItem.callAction(GuiItem.java:178) ~[BellClaims-0.4.4.jar:?]
        at BellClaims-0.4.4.jar/com.github.stefvanschie.inventoryframework.pane.StaticPane.click(StaticPane.java:238) ~[BellClaims-0.4.4.jar:?]
        at BellClaims-0.4.4.jar/com.github.stefvanschie.inventoryframework.gui.InventoryComponent.click(InventoryComponent.java:205) ~[BellClaims-0.4.4.jar:?]
        at BellClaims-0.4.4.jar/com.github.stefvanschie.inventoryframework.gui.type.ChestGui.click(ChestGui.java:200) ~[BellClaims-0.4.4.jar:?]
        at BellClaims-0.4.4.jar/com.github.stefvanschie.inventoryframework.gui.GuiListener.onInventoryClick(GuiListener.java:83) ~[BellClaims-0.4.4.jar:?]
        at com.destroystokyo.paper.event.executor.asm.generated.GeneratedEventExecutor850.execute(Unknown Source) ~[?:?]
        at org.bukkit.plugin.EventExecutor$2.execute(EventExecutor.java:77) ~[paper-mojangapi-1.21-R0.1-SNAPSHOT.jar:?]
        at co.aikar.timings.TimedEventExecutor.execute(TimedEventExecutor.java:80) ~[paper-mojangapi-1.21-R0.1-SNAPSHOT.jar:1.21-2284-4e55e26]
        at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:70) ~[paper-mojangapi-1.21-R0.1-SNAPSHOT.jar:?]
        at io.papermc.paper.plugin.manager.PaperEventManager.callEvent(PaperEventManager.java:54) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at io.papermc.paper.plugin.manager.PaperPluginManagerImpl.callEvent(PaperPluginManagerImpl.java:131) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:628) ~[paper-mojangapi-1.21-R0.1-SNAPSHOT.jar:?]
        at net.minecraft.server.network.ServerGamePacketListenerImpl.handleContainerClick(ServerGamePacketListenerImpl.java:3204) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.network.protocol.game.ServerboundContainerClickPacket.handle(ServerboundContainerClickPacket.java:69) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.network.protocol.game.ServerboundContainerClickPacket.handle(ServerboundContainerClickPacket.java:33) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.network.protocol.PacketUtils.lambda$ensureRunningOnSameThread$0(PacketUtils.java:56) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.server.TickTask.run(TickTask.java:18) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.util.thread.BlockableEventLoop.doRunTask(BlockableEventLoop.java:151) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.util.thread.ReentrantBlockableEventLoop.doRunTask(ReentrantBlockableEventLoop.java:24) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.server.MinecraftServer.doRunTask(MinecraftServer.java:1573) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.server.MinecraftServer.doRunTask(MinecraftServer.java:201) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.util.thread.BlockableEventLoop.pollTask(BlockableEventLoop.java:125) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.server.MinecraftServer.pollTaskInternal(MinecraftServer.java:1550) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.server.MinecraftServer.pollTask(MinecraftServer.java:1543) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.util.thread.BlockableEventLoop.runAllTasks(BlockableEventLoop.java:114) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.server.MinecraftServer.tickServer(MinecraftServer.java:1665) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1334) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:333) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at java.base/java.lang.Thread.run(Thread.java:1583) ~[?:?]`

Puprur 1.21 (1.21.0)
ChatGPT suggests it has to do with being unable to open anvil interface and suggests that there is no quick fix from my side

ok, older 0.4.3 worked, but then using claim tool i get another error

[23:34:53 ERROR]: Could not pass event PlayerInteractEvent to BellClaims v0.4.3
java.lang.NoClassDefFoundError: Could not initialize class dev.mizarc.bellclaims.utils.NonFullBlockFilteringKt
        at BellClaims-0.4.3.jar/dev.mizarc.bellclaims.infrastructure.services.VisualisationServiceBukkit.findSolidBlock(VisualisationServiceBukkit.kt:181) ~[BellClaims-0.4.3.jar:?]
        at BellClaims-0.4.3.jar/dev.mizarc.bellclaims.infrastructure.services.VisualisationServiceBukkit.get3DPositions(VisualisationServiceBukkit.kt:158) ~[BellClaims-0.4.3.jar:?]
        at BellClaims-0.4.3.jar/dev.mizarc.bellclaims.infrastructure.services.VisualisationServiceBukkit.get3DOuterBorders(VisualisationServiceBukkit.kt:119) ~[BellClaims-0.4.3.jar:?]
        at BellClaims-0.4.3.jar/dev.mizarc.bellclaims.infrastructure.services.VisualisationServiceBukkit.displayComplete(VisualisationServiceBukkit.kt:50) ~[BellClaims-0.4.3.jar:?]
        at BellClaims-0.4.3.jar/dev.mizarc.bellclaims.application.actions.player.visualisation.DisplayVisualisation.displayComplete(DisplayVisualisation.kt:88) ~[BellClaims-0.4.3.jar:?]
        at BellClaims-0.4.3.jar/dev.mizarc.bellclaims.application.actions.player.visualisation.DisplayVisualisation.execute(DisplayVisualisation.kt:50) ~[BellClaims-0.4.3.jar:?]
        at BellClaims-0.4.3.jar/dev.mizarc.bellclaims.interaction.listeners.EditToolListener.selectNewCorner(EditToolListener.kt:200) ~[BellClaims-0.4.3.jar:?]
        at BellClaims-0.4.3.jar/dev.mizarc.bellclaims.interaction.listeners.EditToolListener.onUseClaimTool(EditToolListener.kt:128) ~[BellClaims-0.4.3.jar:?]
        at com.destroystokyo.paper.event.executor.asm.generated.GeneratedEventExecutor582.execute(Unknown Source) ~[?:?]
        at org.bukkit.plugin.EventExecutor$2.execute(EventExecutor.java:77) ~[paper-mojangapi-1.21-R0.1-SNAPSHOT.jar:?]
        at co.aikar.timings.TimedEventExecutor.execute(TimedEventExecutor.java:80) ~[paper-mojangapi-1.21-R0.1-SNAPSHOT.jar:1.21-2284-4e55e26]
        at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:70) ~[paper-mojangapi-1.21-R0.1-SNAPSHOT.jar:?]
        at io.papermc.paper.plugin.manager.PaperEventManager.callEvent(PaperEventManager.java:54) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at io.papermc.paper.plugin.manager.PaperPluginManagerImpl.callEvent(PaperPluginManagerImpl.java:131) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:628) ~[paper-mojangapi-1.21-R0.1-SNAPSHOT.jar:?]
        at org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerInteractEvent(CraftEventFactory.java:602) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.server.level.ServerPlayerGameMode.useItemOn(ServerPlayerGameMode.java:538) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.server.network.ServerGamePacketListenerImpl.handleUseItemOn(ServerGamePacketListenerImpl.java:1949) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.network.protocol.game.ServerboundUseItemOnPacket.handle(ServerboundUseItemOnPacket.java:44) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.network.protocol.game.ServerboundUseItemOnPacket.handle(ServerboundUseItemOnPacket.java:11) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.network.protocol.PacketUtils.lambda$ensureRunningOnSameThread$0(PacketUtils.java:56) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.server.TickTask.run(TickTask.java:18) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.util.thread.BlockableEventLoop.doRunTask(BlockableEventLoop.java:151) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.util.thread.ReentrantBlockableEventLoop.doRunTask(ReentrantBlockableEventLoop.java:24) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.server.MinecraftServer.doRunTask(MinecraftServer.java:1573) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.server.MinecraftServer.doRunTask(MinecraftServer.java:201) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.util.thread.BlockableEventLoop.pollTask(BlockableEventLoop.java:125) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.server.MinecraftServer.pollTaskInternal(MinecraftServer.java:1550) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.server.MinecraftServer.pollTask(MinecraftServer.java:1543) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.util.thread.BlockableEventLoop.managedBlock(BlockableEventLoop.java:135) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.server.MinecraftServer.managedBlock(MinecraftServer.java:1502) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.server.MinecraftServer.waitUntilNextTick(MinecraftServer.java:1509) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1354) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:333) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at java.base/java.lang.Thread.run(Thread.java:1583) ~[?:?]
Caused by: java.lang.ExceptionInInitializerError: Exception java.lang.NoSuchFieldError: Class org.bukkit.Material does not have member field 'org.bukkit.Material LEAF_LITTER' [in thread "Server thread"]
        at BellClaims-0.4.3.jar/dev.mizarc.bellclaims.utils.NonFullBlockFilteringKt.<clinit>(NonFullBlockFiltering.kt:436) ~[BellClaims-0.4.3.jar:?]
        at BellClaims-0.4.3.jar/dev.mizarc.bellclaims.infrastructure.services.VisualisationServiceBukkit.findSolidBlock(VisualisationServiceBukkit.kt:181) ~[BellClaims-0.4.3.jar:?]
        at BellClaims-0.4.3.jar/dev.mizarc.bellclaims.infrastructure.services.VisualisationServiceBukkit.get3DPositions(VisualisationServiceBukkit.kt:158) ~[BellClaims-0.4.3.jar:?]
        at BellClaims-0.4.3.jar/dev.mizarc.bellclaims.infrastructure.services.VisualisationServiceBukkit.get3DOuterBorders(VisualisationServiceBukkit.kt:119) ~[BellClaims-0.4.3.jar:?]
        at BellClaims-0.4.3.jar/dev.mizarc.bellclaims.infrastructure.services.VisualisationServiceBukkit.displayComplete(VisualisationServiceBukkit.kt:50) ~[BellClaims-0.4.3.jar:?]
        at BellClaims-0.4.3.jar/dev.mizarc.bellclaims.application.actions.player.visualisation.DisplayVisualisation.displayComplete(DisplayVisualisation.kt:88) ~[BellClaims-0.4.3.jar:?]
        at BellClaims-0.4.3.jar/dev.mizarc.bellclaims.application.actions.player.visualisation.DisplayVisualisation.execute(DisplayVisualisation.kt:50) ~[BellClaims-0.4.3.jar:?]
        at BellClaims-0.4.3.jar/dev.mizarc.bellclaims.application.actions.player.tool.SyncToolVisualization.execute(SyncToolVisualization.kt:23) ~[BellClaims-0.4.3.jar:?]
        at BellClaims-0.4.3.jar/dev.mizarc.bellclaims.interaction.listeners.EditToolVisualisingListener.handleAutoVisualisation(EditToolVisualisingListener.kt:66) ~[BellClaims-0.4.3.jar:?]
        at BellClaims-0.4.3.jar/dev.mizarc.bellclaims.interaction.listeners.EditToolVisualisingListener.onHoldClaimTool$lambda$0(EditToolVisualisingListener.kt:28) ~[BellClaims-0.4.3.jar:?]
        at org.bukkit.craftbukkit.scheduler.CraftTask.run(CraftTask.java:86) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at org.bukkit.craftbukkit.scheduler.CraftScheduler.mainThreadHeartbeat(CraftScheduler.java:475) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.server.MinecraftServer.tickChildren(MinecraftServer.java:1762) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.server.dedicated.DedicatedServer.tickChildren(DedicatedServer.java:509) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.server.MinecraftServer.tickServer(MinecraftServer.java:1634) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1334) ~[purpur-1.21.jar:1.21-2284-4e55e26]
        ... 2 more
stop

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,7 @@ The format is based on Keep a Changelog, and this project adheres to Semantic Ve
 ## [0.5.0]
 
 ### Added
+- Support for Minecraft 1.21.11
 - Action bar popup text on entering and exiting claim.
 - New flag for villagers' ability to use doors.
 - Auto claim visualisation refreshes when moving between chunks, joining the server, and switching dimensions.
@@ -23,6 +24,7 @@ The format is based on Keep a Changelog, and this project adheres to Semantic Ve
 - Player permissions can now be modified when the player is offline.
 - The button to Select all on player permissions works as intended.
 - Falling blocks don't fall when mob flag is active.
+- Versions 1.21.0-1.21.3 not working due to inventory framework.
 
 ## [0.4.4]
 

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -65,7 +65,8 @@ dependencies {
     implementation("com.zaxxer:HikariCP:5.1.0")
     implementation("co.aikar:acf-paper:0.5.1-SNAPSHOT")
     implementation("co.aikar:idb-core:1.0.0-SNAPSHOT")
-    implementation("com.github.mizarc:IF:0.11.4-d")
+    //implementation("com.github.mizarc:IF:0.11.4-d")
+    implementation("com.github.stefvanschie.inventoryframework:IF:0.11.6")
     implementation("io.insert-koin:koin-core:4.0.2")
     compileOnly("com.github.MilkBowl:VaultAPI:1.7") {
         exclude(group = "org.bukkit", module = "bukkit")
@@ -91,6 +92,6 @@ tasks.register<Copy>("deploy") {
     from(layout.buildDirectory.dir("libs"))
     println("Target deployment path: ${getProperty("plugin.server.path")}")
     into(getProperty("plugin.server.path"))
-    rename { fileName -> "${rootProject.name}-${version}.jar" }
+    rename { _ -> "${rootProject.name}-${version}.jar" }
     duplicatesStrategy = DuplicatesStrategy.INCLUDE
 }

--- a/src/main/kotlin/dev/mizarc/bellclaims/BellClaims.kt
+++ b/src/main/kotlin/dev/mizarc/bellclaims/BellClaims.kt
@@ -191,8 +191,6 @@ class BellClaims : JavaPlugin() {
             Class.forName("io.papermc.paper.event.player.PlayerClientLoadedWorldEvent")
             server.pluginManager.registerEvents(EditToolVisualisingListenerExtra(editToolVisualisingListener), this)
         } catch (e: ClassNotFoundException) {
-            // We are on an older version; PlayerClientLoadedWorldEvent logic is skipped.
-            // The fallback in EditToolVisualisingListener (the 5s timeout) will handle cleanup.
             logger.info("Skipping modern client-load events: Server version is older than 1.21.4.")
         }
 

--- a/src/main/kotlin/dev/mizarc/bellclaims/BellClaims.kt
+++ b/src/main/kotlin/dev/mizarc/bellclaims/BellClaims.kt
@@ -182,7 +182,20 @@ class BellClaims : JavaPlugin() {
         server.pluginManager.registerEvents(CloseInventoryListener(), this)
         server.pluginManager.registerEvents(EditToolListener(), this)
         server.pluginManager.registerEvents(ClaimEnterListener(), this)
-        server.pluginManager.registerEvents(EditToolVisualisingListener(this), this)
+
+        val editToolVisualisingListener = EditToolVisualisingListener(this)
+        server.pluginManager.registerEvents(editToolVisualisingListener, this)
+
+        // Try to register the modern "Extra" listener only if the class exists
+        try {
+            Class.forName("io.papermc.paper.event.player.PlayerClientLoadedWorldEvent")
+            server.pluginManager.registerEvents(EditToolVisualisingListenerExtra(editToolVisualisingListener), this)
+        } catch (e: ClassNotFoundException) {
+            // We are on an older version; PlayerClientLoadedWorldEvent logic is skipped.
+            // The fallback in EditToolVisualisingListener (the 5s timeout) will handle cleanup.
+            logger.info("Skipping modern client-load events: Server version is older than 1.21.4.")
+        }
+
         server.pluginManager.registerEvents(HarvestReplantListener(), this)
         server.pluginManager.registerEvents(MoveToolListener(), this)
         server.pluginManager.registerEvents(PartitionUpdateListener(), this)

--- a/src/main/kotlin/dev/mizarc/bellclaims/interaction/listeners/EditToolVisualisingListener.kt
+++ b/src/main/kotlin/dev/mizarc/bellclaims/interaction/listeners/EditToolVisualisingListener.kt
@@ -25,21 +25,11 @@ class EditToolVisualisingListener(private val plugin: JavaPlugin): Listener, Koi
     private val syncToolVisualization: SyncToolVisualization by inject()
 
     // Track players who've just loaded in so we can ignore spurious inventory events until they're fully in-game
-    private val initialisingPlayers: MutableSet<UUID> = ConcurrentHashMap.newKeySet()
+    internal val initialisingPlayers: MutableSet<UUID> = ConcurrentHashMap.newKeySet()
 
     @EventHandler
     fun onPlayerPreLogin(event: AsyncPlayerPreLoginEvent) {
         initialisingPlayers.add(event.uniqueId)
-    }
-
-    /**
-     * Triggers when the player joins the server.
-     */
-    @EventHandler
-    fun onPlayerJoin(event: PlayerClientLoadedWorldEvent) {
-        val player = event.player
-        handleAutoVisualisation(player)
-        initialisingPlayers.remove(player.uniqueId)
     }
 
     /**
@@ -102,7 +92,7 @@ class EditToolVisualisingListener(private val plugin: JavaPlugin): Listener, Koi
      * Visualise if the player isn't already holding the claim tool (e.g. swapping hands)
      * Also update the tracked set depending on whether the player holds the tool after the sync.
      */
-    private fun handleAutoVisualisation(player: Player) {
+    internal fun handleAutoVisualisation(player: Player) {
         val playerId = player.uniqueId
         val mainHand = player.inventory.itemInMainHand
         val offHand = player.inventory.itemInOffHand

--- a/src/main/kotlin/dev/mizarc/bellclaims/interaction/listeners/EditToolVisualisingListenerExtra.kt
+++ b/src/main/kotlin/dev/mizarc/bellclaims/interaction/listeners/EditToolVisualisingListenerExtra.kt
@@ -1,0 +1,19 @@
+package dev.mizarc.bellclaims.interaction.listeners
+
+import io.papermc.paper.event.player.PlayerClientLoadedWorldEvent
+import org.bukkit.event.EventHandler
+import org.bukkit.event.Listener
+import org.koin.core.component.KoinComponent
+
+class EditToolVisualisingListenerExtra(private val mainListener: EditToolVisualisingListener): Listener, KoinComponent {
+
+    /**
+     * Triggers when the player joins the server.
+     */
+    @EventHandler
+    fun onPlayerWorldSwitch(event: PlayerClientLoadedWorldEvent) {
+        val player = event.player
+        mainListener.handleAutoVisualisation(player)
+        mainListener.initialisingPlayers.remove(player.uniqueId)
+    }
+}

